File: implementations/java/client/src/main/java/com/theshuai/tunnelclient/peer/PeerDataFrameCodec.java
package com.theshuai.tunnelclient.peer;

import javax.crypto.Cipher;
import javax.crypto.Mac;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;

final class PeerDataFrameCodec {
    static final int MAGIC = 0x53504D32; // SPM2
    private static final int NONCE_BYTES = 12;
    private static final int TAG_BITS = 128;
    private static final int TAG_BYTES = TAG_BITS / Byte.SIZE;
    static final int HEADER_BYTES = Integer.BYTES + Long.BYTES * 2;
    private static final int MIN_BYTES = HEADER_BYTES + TAG_BYTES;
    private static final int MAX_BYTES = 65_535;
    private static final byte[] INFO_PREFIX = "shuai-peer-mesh/spm2/aes-gcm\n"
            .getBytes(StandardCharsets.US_ASCII);
    private static final byte[] EMPTY = new byte[0];
    private static final ThreadLocal<Cipher> ENCRYPT_CIPHER = ThreadLocal.withInitial(PeerDataFrameCodec::newCipher);
    private static final ThreadLocal<Cipher> DECRYPT_CIPHER = ThreadLocal.withInitial(PeerDataFrameCodec::newCipher);

    private PeerDataFrameCodec() {
    }

    /**
     * 派生单向 traffic key。方向由 {@code fromClientId -> toClientId} 决定，
     * {@code senderKeyEpoch} 是**发送方**本次运行实例的随机 epoch。
     *
     * <p>epoch 是 AES-GCM nonce 唯一性的必要前提：sessionId/token 在服务端 TTL 内会被复用，
     * X25519 密钥又持久化在磁盘上，而 sequence 在进程重启后从 1 重新开始。没有 epoch 时，
     * 客户端重启会在同一 key 下重放同一段 nonce 空间，直接摧毁 GCM 的认证性。
     */
    static TrafficKey trafficKey(SecretKeySpec sessionKey,
                                 long sessionId,
                                 long fromClientId,
                                 long toClientId,
                                 String senderKeyEpoch) {
        if (sessionKey == null || sessionId <= 0 || fromClientId <= 0 || toClientId <= 0
                || fromClientId == toClientId) {
            throw new IllegalArgumentException("SPM2 traffic key requires a valid session and direction");
        }
        if (senderKeyEpoch == null || senderKeyEpoch.isBlank()) {
            throw new IllegalArgumentException("SPM2 traffic key requires the sender key epoch");
        }
        try {
            byte[] sessionBytes = ByteBuffer.allocate(Long.BYTES).putLong(sessionId).array();
            Mac extract = Mac.getInstance("HmacSHA256");
            extract.init(new SecretKeySpec(sessionBytes, "HmacSHA256"));
            byte[] prk = extract.doFinal(sessionKey.getEncoded());

            byte[] suffix = (sessionId + "\n" + fromClientId + "\n" + toClientId + "\n" + senderKeyEpoch)
                    .getBytes(StandardCharsets.US_ASCII);
            byte[] info = new byte[INFO_PREFIX.length + suffix.length];
            System.arraycopy(INFO_PREFIX, 0, info, 0, INFO_PREFIX.length);
            System.arraycopy(suffix, 0, info, INFO_PREFIX.length, suffix.length);

            Mac expand = Mac.getInstance("HmacSHA256");
            expand.init(new SecretKeySpec(prk, "HmacSHA256"));
            byte[] firstInput = new byte[info.length + 1];
            System.arraycopy(info, 0, firstInput, 0, info.length);
            firstInput[firstInput.length - 1] = 1;
            byte[] key = expand.doFinal(firstInput);

            expand.reset();
            byte[] secondInput = new byte[key.length + info.length + 1];
            System.arraycopy(key, 0, secondInput, 0, key.length);
            System.arraycopy(info, 0, secondInput, key.length, info.length);
            secondInput[secondInput.length - 1] = 2;
            byte[] nonceMaterial = expand.doFinal(secondInput);
            return new TrafficKey(new SecretKeySpec(key, "AES"), ByteBuffer.wrap(nonceMaterial).getInt());
        } catch (Exception e) {
            throw new IllegalStateException("derive SPM2 traffic key failed", e);
        }
    }

    static byte[] encode(TrafficKey trafficKey,
                         long sessionId,
                         long sequence,
                         byte[] plaintext) {
        byte[] input = plaintext == null ? EMPTY : plaintext;
        return encode(trafficKey, sessionId, sequence, input, 0, input.length);
    }

    static byte[] encode(TrafficKey trafficKey,
                         long sessionId,
                         long sequence,
                         byte[] plaintext,
                         int offset,
                         int length) {
        if (trafficKey == null || sessionId <= 0 || sequence <= 0) {
            throw new IllegalArgumentException("invalid SPM2 key, session, or sequence");
        }
        byte[] input = plaintext == null ? EMPTY : plaintext;
        checkInputRange(input, offset, length);
        if (HEADER_BYTES + length + TAG_BYTES > MAX_BYTES) {
            throw new IllegalArgumentException("SPM2 peer data frame is too large");
        }
        try {
            byte[] frame = new byte[HEADER_BYTES + length + TAG_BYTES];
            ByteBuffer header = ByteBuffer.wrap(frame);
            header.putInt(MAGIC);
            header.putLong(sessionId);
            header.putLong(sequence);
            byte[] nonce = nonce(trafficKey.noncePrefix(), sequence);

            Cipher cipher = ENCRYPT_CIPHER.get();
            cipher.init(Cipher.ENCRYPT_MODE, trafficKey.key(), new GCMParameterSpec(TAG_BITS, nonce));
            cipher.updateAAD(frame, 0, HEADER_BYTES);
            int written = cipher.doFinal(input, offset, length, frame, HEADER_BYTES);
            if (written != length + TAG_BYTES) {
                throw new IllegalStateException("unexpected SPM2 AES-GCM output length: " + written);
            }
            return frame;
        } catch (Exception e) {
            throw new IllegalStateException("encode SPM2 peer data frame failed: " + e.getMessage(), e);
        }
    }

    static PeerDataFrame decode(TrafficKey trafficKey,
                                byte[] packet,
                                long expectedSessionId) {
        try {
            if (trafficKey == null || packet == null || packet.length < MIN_BYTES || packet.length > MAX_BYTES) {
                return null;
            }
            ByteBuffer header = ByteBuffer.wrap(packet);
            if (header.getInt() != MAGIC) {
                return null;
            }
            long sessionId = header.getLong();
            long sequence = header.getLong();
            if (sessionId != expectedSessionId || sequence <= 0) {
                return null;
            }
            byte[] nonce = nonce(trafficKey.noncePrefix(), sequence);
            Cipher cipher = DECRYPT_CIPHER.get();
            cipher.init(Cipher.DECRYPT_MODE, trafficKey.key(), new GCMParameterSpec(TAG_BITS, nonce));
            cipher.updateAAD(packet, 0, HEADER_BYTES);
            byte[] plaintext = cipher.doFinal(packet, HEADER_BYTES, packet.length - HEADER_BYTES);
            return new PeerDataFrame(sessionId, sequence, plaintext);
        } catch (Exception e) {
            return null;
        }
    }

    static Long sessionId(byte[] packet) {
        return packet == null ? null : sessionId(packet, 0, packet.length);
    }

    static Long sessionId(byte[] packet, int offset, int length) {
        if (packet == null
                || offset < 0
                || length < MIN_BYTES
                || length > MAX_BYTES
                || offset > packet.length - length) {
            return null;
        }
        ByteBuffer header = ByteBuffer.wrap(packet, offset, length);
        return header.getInt() == MAGIC ? header.getLong() : null;
    }

    static boolean looksLikeDataFrame(byte[] packet, int offset, int length) {
        return packet != null
                && offset >= 0
                && length >= Integer.BYTES
                && offset <= packet.length - length
                && ByteBuffer.wrap(packet, offset, Integer.BYTES).getInt() == MAGIC;
    }

    private static Cipher newCipher() {
        try {
            return Cipher.getInstance("AES/GCM/NoPadding");
        } catch (Exception e) {
            throw new IllegalStateException("AES-GCM is unavailable", e);
        }
    }

    private static void checkInputRange(byte[] input, int offset, int length) {
        if (offset < 0 || length < 0 || offset > input.length - length) {
            throw new IndexOutOfBoundsException("invalid peer data frame payload range");
        }
    }

    private static byte[] nonce(int prefix, long sequence) {
        return ByteBuffer.allocate(NONCE_BYTES).putInt(prefix).putLong(sequence).array();
    }

    record TrafficKey(SecretKeySpec key, int noncePrefix) {
    }
}

File: implementations/csharp/client/src/Specus.Client/PeerMesh/PeerDataFrameCodec.cs
using System.Buffers.Binary;
using System.Security.Cryptography;
using System.Text;

namespace Specus.Client.PeerMesh;

internal static class PeerDataFrameCodec
{
    public const uint Magic = 0x53504d32; // SPM2
    private const int NonceBytes = 12;
    private const int TagBytes = 16;
    internal const int HeaderBytes = 4 + 8 * 2;
    private const int MinFrameBytes = HeaderBytes + TagBytes;
    private const int MaxFrameBytes = 65535;

    public static byte[] Encode(
        byte[] aesKey,
        long sessionId,
        long fromClientId,
        long toClientId,
        string senderKeyEpoch,
        long sequence,
        ReadOnlySpan<byte> payload)
    {
        using var codec = CreateTrafficCodec(aesKey, sessionId, fromClientId, toClientId, senderKeyEpoch);
        return codec.Encode(sessionId, sequence, payload);
    }

    public static PeerDataFrame Decode(
        byte[] aesKey,
        long expectedFromClientId,
        long expectedToClientId,
        string senderKeyEpoch,
        ReadOnlySpan<byte> frame)
    {
        if (aesKey.Length != 32
            || frame.Length < MinFrameBytes
            || frame.Length > MaxFrameBytes
            || BinaryPrimitives.ReadUInt32BigEndian(frame[..4]) != Magic)
        {
            throw new CryptographicException("invalid SPM2 peer data frame");
        }
        var sessionId = (long)BinaryPrimitives.ReadUInt64BigEndian(frame.Slice(4, 8));
        var sequence = (long)BinaryPrimitives.ReadUInt64BigEndian(frame.Slice(12, 8));
        if (sessionId <= 0 || expectedFromClientId <= 0 || expectedToClientId <= 0
            || expectedFromClientId == expectedToClientId || sequence <= 0)
        {
            throw new CryptographicException("invalid SPM2 session, direction, or sequence");
        }

        using var codec = CreateTrafficCodec(
            aesKey, sessionId, expectedFromClientId, expectedToClientId, senderKeyEpoch);
        return codec.Decode(sessionId, frame);
    }

    public static bool LooksLikeDataFrame(ReadOnlySpan<byte> value) =>
        value.Length >= 4 && BinaryPrimitives.ReadUInt32BigEndian(value[..4]) == Magic;

    public static long? SessionId(ReadOnlySpan<byte> value) =>
        value.Length is >= MinFrameBytes and <= MaxFrameBytes && LooksLikeDataFrame(value)
            ? (long)BinaryPrimitives.ReadUInt64BigEndian(value.Slice(4, 8))
            : null;

    internal static TrafficCodec CreateTrafficCodec(
        byte[] aesKey,
        long sessionId,
        long fromClientId,
        long toClientId,
        string senderKeyEpoch)
    {
        if (aesKey.Length != 32 || sessionId <= 0 || fromClientId <= 0 || toClientId <= 0
            || fromClientId == toClientId)
        {
            throw new CryptographicException("invalid SPM2 key, session, or direction");
        }
        if (string.IsNullOrWhiteSpace(senderKeyEpoch))
        {
            throw new CryptographicException("SPM2 traffic key requires the sender key epoch");
        }
        var (trafficKey, noncePrefix) = DeriveTrafficMaterial(
            aesKey, sessionId, fromClientId, toClientId, senderKeyEpoch);
        return new TrafficCodec(trafficKey, noncePrefix);
    }

    /// <summary>
    /// Derives the one-way traffic key. <paramref name="senderKeyEpoch"/> is the sender's
    /// per-process random epoch and is mandatory: sessionId/token are reused within the server
    /// session TTL and X25519 keys are persisted on disk, so without a fresh epoch a client
    /// restart would replay the same nonce space under the same AES-GCM key.
    /// </summary>
    private static (byte[] Key, uint NoncePrefix) DeriveTrafficMaterial(
        byte[] aesKey,
        long sessionId,
        long fromClientId,
        long toClientId,
        string senderKeyEpoch)
    {
        Span<byte> salt = stackalloc byte[8];
        BinaryPrimitives.WriteUInt64BigEndian(salt, (ulong)sessionId);
        var prk = HMACSHA256.HashData(salt, aesKey);
        var infoText = $"specus-peer-mesh/spm2/aes-gcm\n{sessionId}\n{fromClientId}\n{toClientId}\n{senderKeyEpoch}";
        var info = Encoding.ASCII.GetBytes(infoText);
        var material = HkdfExpand(prk, info, 36);
        return (material[..32], BinaryPrimitives.ReadUInt32BigEndian(material.AsSpan(32, 4)));
    }

    private static byte[] HkdfExpand(byte[] prk, byte[] info, int length)
    {
        var result = new byte[length];
        var previous = Array.Empty<byte>();
        var offset = 0;
        byte counter = 1;
        while (offset < result.Length)
        {
            var input = new byte[previous.Length + info.Length + 1];
            previous.CopyTo(input, 0);
            info.CopyTo(input, previous.Length);
            input[^1] = counter++;
            previous = HMACSHA256.HashData(prk, input);
            var count = Math.Min(previous.Length, result.Length - offset);
            previous.AsSpan(0, count).CopyTo(result.AsSpan(offset));
            offset += count;
        }
        return result;
    }

    private static void WriteNonce(Span<byte> nonce, uint prefix, ulong sequence)
    {
        BinaryPrimitives.WriteUInt32BigEndian(nonce[..4], prefix);
        BinaryPrimitives.WriteUInt64BigEndian(nonce[4..], sequence);
    }

    internal sealed class TrafficCodec(byte[] trafficKey, uint noncePrefix) : IDisposable
    {
        private readonly AesGcm _gcm = new(trafficKey, TagBytes);
        private readonly object _sync = new();

        internal byte[] Encode(long sessionId, long sequence, ReadOnlySpan<byte> payload)
        {
            if (sessionId <= 0 || sequence <= 0)
            {
                throw new CryptographicException("invalid SPM2 session or sequence");
            }
            if (HeaderBytes + payload.Length + TagBytes > MaxFrameBytes)
            {
                throw new CryptographicException("SPM2 peer data frame is too large");
            }
            var frame = new byte[HeaderBytes + payload.Length + TagBytes];
            BinaryPrimitives.WriteUInt32BigEndian(frame.AsSpan(0, 4), Magic);
            BinaryPrimitives.WriteUInt64BigEndian(frame.AsSpan(4, 8), (ulong)sessionId);
            BinaryPrimitives.WriteUInt64BigEndian(frame.AsSpan(12, 8), (ulong)sequence);
            Span<byte> nonce = stackalloc byte[NonceBytes];
            WriteNonce(nonce, noncePrefix, (ulong)sequence);
            lock (_sync)
            {
                _gcm.Encrypt(
                    nonce,
                    payload,
                    frame.AsSpan(HeaderBytes, payload.Length),
                    frame.AsSpan(HeaderBytes + payload.Length, TagBytes),
                    frame.AsSpan(0, HeaderBytes));
            }
            return frame;
        }

        internal PeerDataFrame Decode(long expectedSessionId, ReadOnlySpan<byte> frame)
        {
            if (frame.Length < MinFrameBytes
                || frame.Length > MaxFrameBytes
                || BinaryPrimitives.ReadUInt32BigEndian(frame[..4]) != Magic)
            {
                throw new CryptographicException("invalid SPM2 peer data frame");
            }
            var sessionId = (long)BinaryPrimitives.ReadUInt64BigEndian(frame.Slice(4, 8));
            var sequence = (long)BinaryPrimitives.ReadUInt64BigEndian(frame.Slice(12, 8));
            if (sessionId <= 0 || sessionId != expectedSessionId || sequence <= 0)
            {
                throw new CryptographicException("invalid SPM2 session or sequence");
            }
            Span<byte> nonce = stackalloc byte[NonceBytes];
            WriteNonce(nonce, noncePrefix, (ulong)sequence);
            var payloadLength = frame.Length - HeaderBytes - TagBytes;
            var payload = new byte[payloadLength];
            lock (_sync)
            {
                _gcm.Decrypt(
                    nonce,
                    frame.Slice(HeaderBytes, payloadLength),
                    frame.Slice(HeaderBytes + payloadLength, TagBytes),
                    payload,
                    frame[..HeaderBytes]);
            }
            return new PeerDataFrame(sessionId, sequence, payload);
        }

        public void Dispose()
        {
            lock (_sync)
            {
                _gcm.Dispose();
            }
        }
    }
}

internal sealed record PeerDataFrame(
    long SessionId,
    long Sequence,
    byte[] Payload);
